### PR TITLE
feat(`tofs`): allow TOFS for master+minion configuration

### PR DESCRIPTION
**Pull request by myii**
* Sourced from master config files for all major versions since `2016.11`:
  - `2016.11`
  - `2017.7`
  - `2018.3`
  - `2019.2`

---

The current [master.d/f_defaults.conf](https://github.com/saltstack-formulas/salt-formula/blob/698085f7580ec7710ea5ca97b42bb773b599c2a2/salt/files/master.d/f_defaults.conf) is limited to settings from `2016.11`.  There have been two more major releases since then and another on the way soon.

This PR starts the process of ensuring that this file works for any master since `2016.11`.  I've done most of the work but I'd appreciate this being reviewed, due to the importance of this configuration as well as some of the complexity in it.

Of particular importance is the new `get_jinja_env_config_block` macro and how it populates the two Jinja environment sections.  Problems here can cause breakages in templates and state files and I've spent a lot of time ensuring this all works.  It has also resulted in uncovering a bug that I've reported in a new [upstream issue](https://github.com/saltstack/salt/issues/51447).  I've applied the workaround for that in this PR.

If this all works well, the same can be done for the [minion.d/f_defaults.conf](https://github.com/saltstack-formulas/salt-formula/blob/698085f7580ec7710ea5ca97b42bb773b599c2a2/salt/files/minion.d/f_defaults.conf) file as well.

For ease of comparison, the four source (master config) files are:

1. https://github.com/saltstack/salt/blob/2016.11/conf/master
1. https://github.com/saltstack/salt/blob/2017.7/conf/master
1. https://github.com/saltstack/salt/blob/2018.3/conf/master
1. https://github.com/saltstack/salt/blob/2019.2/conf/master

### CODE REVIEW

**Comment by myii** [comment]
@aboe76 Another way this could be done is by having separate source master files for each major version.  But that still requires the time to check each setting and the complexity of the Jinja environment parameters cannot be avoided (that I've attempted with the macro `get_jinja_env_config_block`).  I can't see any advantage to doing this separation.

What it really boils down to is whether this should be supported by this formula or not.  If not, we will have to reconsider removing the current `f_defaults.conf` files, since these could end up causing big problems with the latest installations of Salt.

**Comment by getSurreal** [comment]
Would having a template for each major release vs one template with a bunch of if statements be easier to maintain long term.

**Comment by myii** [comment]
@getSurreal Thanks for the feedback.  That's what I've mentioned in the comment just above yours.  Personally, I'm not so sure the `if` statements really bring much of the complexity.  Testing the macros and settings being applied is where it really gets difficult (even worse if across multiple files).  And then having to separate this in the pillar seems even more complex.  Not to mention making modifications across multiple master config files when there are fixes/improvements required (for the common bits, which is most of the file).

**Comment by myii** [comment]
@aboe76 @getSurreal Actually, I've looked into this and I'm convinced that using multiple files would far worse to maintain.

### Only 4 releases need to covered by the file at any one time

https://s.saltstack.com/product-support-lifecycle/:

Salt Open releases|Phase 1 support ends|Phase 2 support ends|Phase 3 support ends|Extended life support ends
-----|-----|-----|-----|-----
Salt Open 2018.3|Oct 31, 2018|Apr 30, 2019|Oct 31, 2019|Oct 31, 2020
Salt Open 2017.7|Jan 31, 2018|Jun 30, 2018|Dec 31, 2018|Dec 31, 2019
Salt Open 2016.11|May 31, 2017|Nov. 30, 2017|May 31, 2018|May 31, 2019

* Roughly 8 months between releases.
* No more than 32 months until extended life support ends.
* 32 / 8 = 4 active major versions at one time.

To assist with this process of maintenance, I propose using the same method as the main docs as a comment for each `if` block, such as:

* versionadded: 2017.7
* versionchanged: 2017.7
* deprecated: 2018.3

That will make it much easier to clear blocks once a major version is no longer supported (without having to check each `if` block to work out if it needs to be removed or not).

### Git similarity between versions is always > 90%

Tested using the similarity algorithm used by git for detecting file renames:

```git
 1 file changed, 91 insertions(+), 33 deletions(-)
 rename 2016.11 => 2017.7 (92%)

 1 file changed, 85 insertions(+), 22 deletions(-)
 rename 2017.7 => 2018.3 (93%)

 1 file changed, 14 insertions(+), 6 deletions(-)
 rename 2018.3 => 2019.2 (99%)
```

* With such a high similarity index, the duplication across 4 active files will be far more difficult to maintain than a single file.

**Comment by getSurreal** [comment]
@myii You make a good case for a single file and I like matching it to the actual salt release cycle.

**Comment by myii** [comment]
@getSurreal Thanks again for your time and feedback.

**Comment by myii** [comment]
> To assist with this process of maintenance, I propose using the same method as the main docs as a comment for each `if` block, such as:
> 
> * versionadded: 2017.7
> * versionchanged: 2017.7
> * deprecated: 2018.3

I've just pushed another commit to show these comments added in after each `{%- if major_svi ... %}` and its respective `{%- else %}` where applicable.

**Comment by aboe76** [comment]
What about a radical change, instead of mimicking ```/etc/salt/minion``` and ```/etc/salt/master``` everytime, only parse / serialize the changes from pillar data into the f_defaults.conf 

because the f_defaults.conf file is specially there so the package system of the OS can update the files in ```/etc/salt/master``` and ```/etc/salt/minion``` and gain new features variables commented for the version. f_defaults.conf should only override those values.

Downside to that aproach is we don't have any documentation on what features could be set....

**Comment by myii** [comment]
@aboe76 above:

> What about a radical change, instead of mimicking `/etc/salt/minion` and `/etc/salt/master` everytime, only parse / serialize the changes from pillar data into the f_defaults.conf

So if I understand you correctly:

* This will get rid off:
    * The need to check for `saltversioninfo` like I'm doing in this PR.
    * All of the commented default values.
    * All of the commented documentation.
* What will still remain:
    * All of the macros.
    * All of the logic.

> because the f_defaults.conf file is specially there so the package system of the OS can update the files in `/etc/salt/master` and `/etc/salt/minion` and gain new features variables commented for the version. f_defaults.conf should only override those values.

That makes sense and saves unnecessary duplication of commented values from the `/etc/salt/master` and `/etc/salt/minion` files.

> Downside to that aproach is we don't have any documentation on what features could be set....

What documentation are you referring to here?  The documentation that is kept in `f_defaults.conf`?

---

While we're here throwing curveballs, I've got another idea to add to the discussion.

#### Make use of a new `svimap.yaml`

Globbing works since `2016.11` so this PR (up to now) could be replaced by an alternative method built upon a structure like this:

```yaml
# -*coding: utf-8 -*-
# vim: ft=yaml

2016.11.*:
  master_f_defaults:
    salt_event_pub_hwm: '20000'
    event_publisher_pub_hwm: '10000'
    external_nodes: 'None'
    renderer: 'yaml_jinja'
    jinja_trim_blocks: 'False'
    jinja_lstrip_blocks: 'False'
2017.7.*:
  master_f_defaults:
    keep_acl_in_token: 'False'
    eauth_acl_module: 'django'
    renderer: 'yaml_jinja'
    jinja_trim_blocks: 'False'
    jinja_lstrip_blocks: 'False'
    gitfs_refspecs:
      - '+refs/heads/*:refs/remotes/origin/*'
      - '+refs/tags/*:refs/tags/*'
2018.3.*:
  master_f_defaults:
    batch_safe_limit: 100
    batch_safe_size: 8
    master_stats: 'False'
    master_stats_event_iter: 60
    key_pass: 'sdb://masterkeyring/key_pass'
    master_sign_pubkey: 'True'
    signing_key_pass: 'sdb://masterkeyring/signing_pass'
    autosign_grains_dir: '/etc/salt/autosign_grains'
    keep_acl_in_token: 'False'
    eauth_acl_module: 'django'
    roster: 'flat'
    ssh_update_roster: 'False'
    utils_dirs: '[]'
    renderer: 'yaml_jinja'
    jinja_env:
      block_start_string: '{{ "{%" }}'
      block_end_string: '{{ "%}" }}'
      variable_start_string: '{{ "{{" }}'
      variable_end_string: '{{ "}}" }}'
      comment_start_string: '{{ "{#" }}'
      comment_end_string: '{{ "#}" }}'
      line_statement_prefix:
      line_comment_prefix:
      trim_blocks: False
      lstrip_blocks: False
      # Using double-quotes until https://github.com/saltstack/salt/issues/51447 resolved
      newline_sequence: '"\n"'
      keep_trailing_newline: False
    jinja_sls_env:
      block_start_string: '{{ "{%" }}'
      block_end_string: '{{ "%}" }}'
      variable_start_string: '{{ "{{" }}'
      variable_end_string: '{{ "}}" }}'
      comment_start_string: '{{ "{#" }}'
      comment_end_string: '{{ "#}" }}'
      line_statement_prefix:
      line_comment_prefix:
      trim_blocks: False
      lstrip_blocks: False
      # Using double-quotes until https://github.com/saltstack/salt/issues/51447 resolved
      newline_sequence: '"\n"'
      keep_trailing_newline: False
    gitfs_refspecs:
      - '+refs/heads/*:refs/remotes/origin/*'
      - '+refs/tags/*:refs/tags/*'
2019.2.*:
  master_f_defaults:
    ipc_mode: 'ipc'
    tcp_master_pub_port: 4510
    tcp_master_pull_port: 4511
    batch_safe_limit: 100
    batch_safe_size: 8
    master_stats: 'False'
    master_stats_event_iter: 60
    key_pass: 'sdb://masterkeyring/key_pass'
    master_sign_pubkey: 'True'
    signing_key_pass: 'sdb://masterkeyring/signing_pass'
    autosign_grains_dir: '/etc/salt/autosign_grains'
    keep_acl_in_token: 'False'
    eauth_acl_module: 'django'
    roster: 'flat'
    ssh_update_roster: 'False'
    utils_dirs: '[]'
    renderer: 'jinja|yaml'
    jinja_env:
      block_start_string: '{{ "{%" }}'
      block_end_string: '{{ "%}" }}'
      variable_start_string: '{{ "{{" }}'
      variable_end_string: '{{ "}}" }}'
      comment_start_string: '{{ "{#" }}'
      comment_end_string: '{{ "#}" }}'
      line_statement_prefix:
      line_comment_prefix:
      trim_blocks: False
      lstrip_blocks: False
      # Using double-quotes until https://github.com/saltstack/salt/issues/51447 resolved
      newline_sequence: '"\n"'
      keep_trailing_newline: False
    jinja_sls_env:
      block_start_string: '{{ "{%" }}'
      block_end_string: '{{ "%}" }}'
      variable_start_string: '{{ "{{" }}'
      variable_end_string: '{{ "}}" }}'
      comment_start_string: '{{ "{#" }}'
      comment_end_string: '{{ "#}" }}'
      line_statement_prefix:
      line_comment_prefix:
      trim_blocks: False
      lstrip_blocks: False
      # Using double-quotes until https://github.com/saltstack/salt/issues/51447 resolved
      newline_sequence: '"\n"'
      keep_trailing_newline: False
    gitfs_refspecs:
      - '+refs/heads/*:refs/remotes/origin/*'
      - '+refs/tags/*:refs/tags/*'
```

That shows all of the changes I have done so far.  Imagine all of the common config settings are represented in `defaults.yaml`.  The above can then be merged on top of the common defaults in the usual manner in `map.jinja`:

```jinja
{% import_yaml "salt/svimap.yaml" as svimap %}

{# merge the svimap #}
{% set svi = salt['grains.filter_by'](svimap, grain='saltversion') or {} %}
{% do salt['defaults.merge'](defaults['salt'], svi) %}
```

The map will then contain all of the config settings specific to the version of Salt.  With some clever use of the pillar, preparing a limited `f_defaults.conf` could be achieved, only containing the settings differing from the default values.

**Comment by aboe76** [comment]
@myii why do you want to keep the deafults per salt-version, they are already programmed inside salt.
the only thing needed to configure it to your liking is to add overriding key in ```f_defaults.conf````

**Comment by myii** [comment]
@aboe76 There are a number of reasons.  Listing them in no particular order:

1. For refactoring and reverting QA:
    1. When refactoring, it's important to ensure that the results before and after are _identical_ -- can't do that without keeping track of the default settings somewhere.
    1. If this major change doesn't work well, we may need to revert back to the current method of using `f_defaults.conf` -- that will require the default settings; and then we have to choose between the ugly Jinja from this PR or to use these maps to provide the defaults instead.
1. Limit the settings from the pillar into `f_defaults.conf`:
    1. Limit to those that work for the specific Salt version on each master and minion -- this can be achieved by comparing if the pillar value is actually in the map for the specific version of Salt.
    1. Limit further to only use those which actually differ from the default settings in the first place -- this can be achieved by direct comparison between the defaults and the pillar values.
1. Use as documentation of all of the possible settings as well as _how_ to actually set them:
    1. `defaults.yaml` and `svimap.yaml` contain all of the possible settings as well as the format they need to be provided in the pillar (e.g. look at the difficulty of providing the `jinja_env` and `jinja_sls_env` settings without a previous example).
    1. Can do this instead of flooding `pillar.example` with numerous examples across all of the possible settings.
1. Every single setting requires some logic or even relies on a macro:
    1. This method will help to trim out unnecessary content as versions become unsupported.
    1. Ultimately, splitting out all of the logic/macros from the monolithic `f_defaults.conf` files would help with maintenance -- and this method would make that much easier.

Reading the above is very dense because I'm summarising each point to avoid writing an essay.  I'm happy to explain if further clarification is needed.

**Comment by aboe76** [comment]
@myii the salt understanding of its configuration can be default  yaml or json structured, so if we can make the pillar yaml input and output the same then there is no need for jinja manipulation. 
However the way salt ingests pillar data we need to serialize it back to yaml. instead of mere passthrough. 

Downside to this approach is:
- the output ordering isn't the same.
- the readability and comments are lost.

**Comment by myii** [comment]
> the salt understanding of its configuration can be default yaml or json structured, so if we can make the pillar yaml input and output the same then there is no need for jinja manipulation.

Now this sounds very interesting if it can be achieved.

> However the way salt ingests pillar data we need to serialize it back to yaml. instead of mere passthrough.

Is there any risk of the yaml coming out differently at the other end due to this process?

> Downside to this approach is:
> 
> * the output ordering isn't the same.
> * the readability and comments are lost.

If the output matches the pillar, then that can be relied on instead.  Ultimately, it's an output file controlled by Salt.  The actual `master` and `minion` files are still available for the documentation.

**Comment by myii** [comment]
@aboe76 If we do follow through with this method, will have to carefully consider what to do about these available settings:

https://github.com/saltstack-formulas/salt-formula/blob/5eee4c61dc5abcca3ac1471acb287d30f9c6177b/pillar.example#L7-L11

**Comment by myii** [comment]
> https://github.com/saltstack-formulas/salt-formula/blob/5eee4c61dc5abcca3ac1471acb287d30f9c6177b/pillar.example#L7-L11

@aboe76 Just realised, we don't just have to worry about these pillar values... rather, there may be installations out there that have _used_ these values and no longer have the default `master` and `minion` configuration files.  So then these files aren't present to refer back to for all of the available settings and comments.

**Comment by aboe76** [comment]
@myii for those setups that have removed the files there is also docs.saltstack....

**Comment by myii** [comment]
@aboe76 After some testing, I'm definite that this idea is not achievable without serious consequences to existing setups.  The following diff is based on using `pillar.example`.  To prepare it, I've had to merge the output from different files, as mentioned in the notes after the diff:

```diff
 --- YAML-only proposal
+++ Current implementation
@@ -1,4 +1,4 @@
-auth.ldap.filter: uid={{ username }}
+auth.ldap.filter: "uid={{ username }}"
 engines:
   - slack:
       aliases:
@@ -16,6 +16,47 @@
 file_roots:
   base:
     - /srv/salt
+      - [Numerous file_roots sourced from other parts of the pillar, e.g. formulas]
+      - ...
+      - ...
+      - ...
+      - ...
+      - ...
+  [Second env]:
+      - [Numerous file_roots sourced from other parts of the pillar, e.g. formulas]
+      - ...
+      - ...
+      - ...
+      - ...
+      - ...
+  [Third env]:
+      - [Numerous file_roots sourced from other parts of the pillar, e.g. formulas]
+      - ...
+      - ...
+      - ...
+      - ...
+      - ...
 fileserver_backend:
   - git
   - s3fs
@@ -42,20 +83,16 @@
 pillar_roots:
   base:
     - /srv/pillar
-reactors:
-  - master/deploy:
+reactor:
+  - 'master/deploy':
     - /srv/salt/reactors/deploy.sls
 rest_tornado:
-  debug: false
-  disable_ssl: false
+  debug: False
+  disable_ssl: False
   port: 8000
   ssl_crt: /etc/pki/api/certs/server.crt
   ssl_key: /etc/pki/api/certs/server.key
-s3.buckets:
-  - bucket1
-  - bucket2
-  - bucket3
-  - bucket4
-s3.key: askdjghsdfjkghWupUjasdflkdfklgjsdfjajkghs
-s3.keyid: GKTADJGHEIQSXMKKRBJ08H
-winrepo_provider: gitpython
+s3.buckets: ["bucket1", "bucket2", "bucket3", "bucket4"]
+s3.key: "askdjghsdfjkghWupUjasdflkdfklgjsdfjajkghs"
+s3.keyid: "GKTADJGHEIQSXMKKRBJ08H"
+winrepo_provider: "gitpython"
```

1. The existing system involves more than managing only `f_defaults.conf`; this pillar also results in managing:
    1. `engine.conf`
    1. `lxc_profiles.conf`
    1. `reactor.conf`
1. Use of single and double quotation marks:
    1. Not maintained in the straight YAML rendering -- specifically set in the templates.
    1. Maybe not in the diff above but some of these are significant and will not work without the quotes.
1. `file_roots` and many other settings sources from more than just the `salt:master` pillar.
    1. All of that logic would have to be reproduced.
1. `reactors` => `reactor` modified by the template.
    1. All of these would have to be reproduced.

Who knows what will happen with all of the other possible settings?

---

Basically, testing a massive change like this would take even more effort than the changes proposed in this PR (which are bad enough).  I see two ways forward for working with the 4 supported versions of Salt at any given time:

1. Continue with the method in this PR, i.e. using Jinja-based conditionals.
1. Use the `svimap.conf` idea, as mentioned in the [comment above](https://github.com/saltstack-formulas/salt-formula/pull/398#issuecomment-459853665).

**Comment by aboe76** [comment]
@myii 4 supported versions?

I thought only three?

Salt Open 2018.3 | Oct 31, 2018 | Apr 30, 2019 | Oct 31, 2019 | Oct 31, 2020
Salt Open 2017.7 | Jan 31, 2018 | Jun 30, 2018 | Dec 31, 2018 | Dec 31, 2019
Salt Open 2016.11 | May 31, 2017 | Nov. 30, 2017 | May 31, 2018 | May 31, 2019

see https://s.saltstack.com/product-support-lifecycle/

But then again Is it necessary to support the same versions as saltstack enterprise support?
I would think that is a massive undertaking....

**Comment by aboe76** [comment]
@myii you are right about the other files, I haven't considered those.
your setup about svimap.yaml looks doable.

**Comment by myii** [comment]
@aboe76 According to the current schedule, there will only be 4 active, supported versions at the same time.  Discontinued versions can be stripped out.  Guessing the dates for the upcoming version `2019.2`:

Salt Open releases|Phase 1 support ends|Phase 2 support ends|Phase 3 support ends|Extended life support ends
-----|-----|-----|-----|-----
Salt Open 2019.2||||Sep 30, 2021
Salt Open 2018.3|Oct 31, 2018|Apr 30, 2019|Oct 31, 2019|Oct 31, 2020
Salt Open 2017.7|Jan 31, 2018|Jun 30, 2018|Dec 31, 2018|Dec 31, 2019
Salt Open 2016.11|May 31, 2017|Nov. 30, 2017|May 31, 2018|May 31, 2019

By the time the next version comes out, `2016.11` will no longer be supported.  And so on.

In terms of complexity, not really.  Quoting from above again:

```git
 1 file changed, 91 insertions(+), 33 deletions(-)
 rename 2016.11 => 2017.7 (92%)

 1 file changed, 85 insertions(+), 22 deletions(-)
 rename 2017.7 => 2018.3 (93%)

 1 file changed, 14 insertions(+), 6 deletions(-)
 rename 2018.3 => 2019.2 (99%)
```

The files change very little per version.  The work I've done for this PR is mostly done.  The only reason it was so complex was due to the new features around `jinja_env` & `jinja_sls_env`.  But regular changes are easy to accommodate.

**Comment by aboe76** [comment]
@myii if it is possible we could cut out the macros and put them in a ```macros.jinja``` file and import that
this will make the f_defaults.conf a bit less daunting.

**Comment by myii** [comment]
@aboe76 OK, I will look further into using `svimap.yaml`.  And I will look at the macros as well, to see what can be done to simplify the file(s).

Thanks for all of your feedback.

**Comment by aboe76** [comment]
@myii no problem you are doing us a great service with all this work.

**Comment by myii** [comment]
I finished the settings over a week ago and didn't get around to pushing them.  Just done that now with 
f4ff0f8.  The baseline of this work is complete but the additional stages remain, viz. looking at `svimap.yaml`, separating the macros and even looking at if this implementation can be readdressed using the TOFS pattern -- if the [TOFS update PR](https://github.com/saltstack-formulas/template-formula/pull/28) is merged, then there may be ways of using the new features to achieve our goals.

**Comment by myii** [comment]
Update: only 3 releases are supported upstream.  So now `2017.7` is the oldest version officially supported.  See [this comment](https://github.com/saltstack-formulas/salt-formula/issues/405#issuecomment-476723926) for more info.

**Comment by aboe76** [comment]
@myii was thinking about this, and we are missing one version, that could come in handy:
```latest```

this will be next version of salt and it gives the opportunity to already test and update the configuration.
what do you think?

**Comment by myii** [comment]
@aboe76 Do you mean for this particular PR or across this whole formula?  Or are we talking about the whole org?!  In any case, that is definitely a idea worth considering.  In an ideal world, I would say yes without thinking.  With the amount of improvements that need to be spread across the existing formulas, I'm hoping for the narrowest scope possible, to avoid all of the maintenance problems that we're getting dragged down by!

**Comment by myii** [comment]
@aboe76 A somewhat radical idea I discussed with @javierbertoli in the Slack group: how about we reset these configuration files based on only `2019.2` for the time being?  Meaning:

* The configuration here is currently for `2016.11`, which is no longer supported.
* This configuration has never been updated for `2017.7` or `2018.3`.

Since no-one is relying on this formula to provide the configuration for these last two versions, let's just start again with `2019.2`, removing all of the Jinja conditionals and then reflect on how to improve this configuration method in the longer-term (e.g. splitting into multiple files including only ever configuring values given in the pillar).  Having only one version supported for now will make this planning much easier.

**Comment by myii** [comment]
More from Slack:

* https://freenode.logbot.info/saltstack-formulas/20190514#c2178174

Essentially, now that we have `semantic-release` and TOFS available, why not just provide the master and minion configuration files as templates and let them be deployed as required at the other end?  Basically, take the YAML and Jinja macros out of the equation entirely.

**Comment by aboe76** [comment]
@myii I'm not against this idea, having only pillar to render `*conf.d/*.conf' files that hold users explicit configuration properties.

**Comment by myii** [comment]
@aboe76 @getSurreal @vutny @daks So with respect to https://github.com/saltstack-formulas/salt-formula/pull/398#issuecomment-492399618, here we have the solution based upon TOFS instead.

1. This PR doesn't introduce any breaking changes -- users will specifically have to opt-in to using TOFS for managing the directory instead.
1. Note, that these are no longer Jinja templates; these are straight reproductions of the files you want "at the other end".
1. Since we're using a simple boolean value, it becomes very simple to switch between the old and new functionality:
    1. This makes it easy for users to prepare a TOFS solution that prepares identical configuration files to the pillar-based method.
    1. But with respect to the non-Jinja point above, one could simply copy the actual live configuration directory to be used as source directory (with TOFS, this could even be done by each minion's `id`).
1. You may notice I've also included the as-yet unused [`minion_config_use_TOFS`](https://github.com/saltstack-formulas/salt-formula/pull/398/files#diff-7414cc5de269d1fe9fd6a7638996ce2aR13) into `pillar.example`; as part of the `WIP`, if this method is agreed upon, it is simple to do the same thing for the minion configuration files as well.

In terms of a deprecation policy:

1. We _could_ simply not have one, leaving the pillar option available -- but I think this is a bad idea since that requires so much maintenance.
1. If we do go ahead, I propose the following rough steps:
    1. Display deprecation warning when configuration is being done by pillar.
    1. Make TOFS the default, making pillar-based opt-in.
    1. Remove pillar-based method, with breaking change commit.

File: pillar.example [hunk 1]
```diff
@@ -46,6 +46,28 @@ salt:
...
+  #   path_prefix: template_alt
+  #   dirs:
+  #     files: files_alt
+  #     default: default_alt
+  #   source_files:
+  #     salt-master:
+  #       - 'alt_master.d'
+
   # salt master config
   master:
```
Comment (myii) [inline pillar.example]: How about putting the value here?

```
  # salt master config
  master_config_use_TOFS: True
  master:
    ...
```

**Comment by aboe76** [comment]
@myii so to my understanding, this proposed change will 
- keep the current pillar -> jinja rendering as is. (default)
- introduce TOFS pattern for 'master.d'/'minion.d' files (with switch to turn it on)

Nice work!!!

**Comment by myii** [comment]
> @myii so to my understanding, this proposed change will
> 
> * keep the current pillar -> jinja rendering as is. (default)
> * introduce TOFS pattern for 'master.d'/'minion.d' files (with switch to turn it on)

@aboe76 That's right, except I haven't covered `minion.d` just yet.  If you're all happy with this method, I'll add that to this PR as well.

**Comment by aboe76** [comment]
@myii yes I'm happy with this method and a good solution.

**Comment by myii** [comment]
@aboe76 @vutny OK, I've finalised this particular PR so that the minion configuration can be provided via. TOFS (opt-in, mirroring same method for the master configuration).  If all is OK, I'd appreciate if this can be merged, so that the next steps can commence.  I'll start a new issue with a checklist based upon the one below:

- [ ] `README`: Update to explain how TOFS can be used instead of pillar for master/minion configuration.
- [ ] `salt.cloud`: (After due consideration) Use the same method to allow TOFS to be used for these configurations: `cloud.maps.d`, `cloud.profiles.d` & `cloud.providers.d`.
- [ ] Deprecation step 1: Display deprecation warning when configuration is being done by pillar.
- [ ] Deprecation step 2: Make TOFS the default, making pillar-based opt-in.
- [ ] Deprecation step 3: Remove pillar-based method, with breaking change commit.

In terms of the unrelated `centos-7` failure, we can by-pass this manually until it is fixed (via. a separate PR).

**Comment by myii** [comment]
Thanks for the review, @noelmcloughlin.

**Comment by myii** [comment]
@vutny Are you satisfied with this implementation?  Can we proceed with the merge?

**Comment by myii** [comment]
@vutny Thanks, do you mind merging?

**Comment by vutny** [comment]
@myii Wanna me to do that?

**Comment by myii** [comment]
@vutny I could always do it myself but I prefer to let others do it when it's a PR authored by me!  If it was a minor change, I would have merged it myself.

**Comment by vutny** [comment]
Okay, let's wait a day if there would be no objections, I'll merge it.

**Comment by myii** [comment]
@vutny I don't think there's anyone left to put up an objection, this PR has been around since the end of January!  But I'm absolutely fine with that, it's not like I'm going to be doing the next stage very soon anyway.

**Comment by aboe76** [comment]
LGTM !

Op wo 12 jun. 2019 09:02 schreef Imran Iqbal <notifications@github.com>:

**Comment by myii** [comment]
Thanks to all involved for your input along the way.
